### PR TITLE
Update Buyback contract

### PR DESCRIPTION
**Pull request by shahthepro**
If you made a contract change, make sure to complete the checklist below before merging it in master.

Refer to our [documentation](https://github.com/OriginProtocol/security) for more details about contract security best practices.

Contract change checklist:
  - [ ] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [ ] Unit tests pass
  - [ ] Slither tests pass with no warning
  - [ ] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

File: contracts/utils/addresses.js
Comment (DanielVF) [inline contracts/utils/addresses.js:127]: I think we are closing down Rinkeby.

File: contracts/contracts/mocks/MockUniswapRouter.sol
Comment (DanielVF) [inline contracts/contracts/mocks/MockUniswapRouter.sol:10]: These were getting commented out?

File: contracts/test/buyback/buyback.js
Comment (DanielVF) [inline contracts/test/buyback/buyback.js:147]: Nice check!

File: contracts/contracts/buyback/Buyback.sol
Comment (DanielVF) [inline contracts/contracts/buyback/Buyback.sol:29]: I don't think we need to swap through OGN, see later comment on the swap path. Since that's the case, we can remove the OGN contract variable?

File: contracts/contracts/buyback/Buyback.sol
Comment (DanielVF) [inline contracts/contracts/buyback/Buyback.sol:93]: We can remove the `onlyVault` and `nonReentrant` modifiers from this, since this function no does no actions. 

This also means the contract does not have to know who the vault is.

File: contracts/contracts/buyback/Buyback.sol
Comment (DanielVF) [inline contracts/contracts/buyback/Buyback.sol:114]: Great to check the preconditions here.

File: contracts/contracts/buyback/Buyback.sol
Comment (DanielVF) [inline contracts/contracts/buyback/Buyback.sol:118]: We don't actually need this check, because the swap will fail if there's not enough OUSD.

File: contracts/contracts/buyback/Buyback.sol
Comment (DanielVF) [inline contracts/contracts/buyback/Buyback.sol:121]: We can make some changes to the path here and save money:

- The usdt/weth9 path can now use the 500 fee since there is lots of liquidity in that pool now, which saves us a lot on fees. 
- Origin removed the OGV/OGN pool and moved liquidity to an ETH/OGV pool. We'll want to use that pool for swapping. This also means we can remove tracking the OGN address from this contract.

**Review by DanielVF — CHANGES_REQUESTED** [review]
We'll also need the deploy file to deploy the contract to mainnet, change where the vault sends funds, and transfer old funds over.

**Comment by shahthepro** [comment]
> We'll also need the deploy file to deploy the contract to mainnet, change where the vault sends funds, and transfer old funds over.

Yep. I'm working on that, added a migration file for the first two (not yet pushed). Will tweak it so that the proposal also includes transferring funds

File: contracts/contracts/buyback/Buyback.sol
Comment (shahthepro) [reply contracts/contracts/buyback/Buyback.sol:96]: Have change it, thanks for the tip :)

File: contracts/contracts/buyback/Buyback.sol
Comment (shahthepro) [reply contracts/contracts/buyback/Buyback.sol]: Yep, removed it

File: contracts/contracts/buyback/Buyback.sol
Comment (shahthepro) [reply contracts/contracts/buyback/Buyback.sol]: Ah, yes. Have changed it, also removed all references to vault entirely

**Comment by shahthepro** [comment]
Tried the newly added migration on mainnet fork with a script to test things. Didn't commit the script, but here is the code and the stdout, in case someone else would like to try it out too:

**stdout**:
```New Buyback 0x4258a6E32Eb33Eed64c3caefe6bcCf53141518dE
New Buyback governor 0x72426BA137DEC62657306b12B1E869d43FeC6eC7
New Buyback RewardsSource 0x7d82E86CF1496f9485a8ea04012afeb3C7489397
Updated trustee 0x4258a6E32Eb33Eed64c3caefe6bcCf53141518dE
Old contract OGN balance 0
Old contract OUSD balance 0
New contract OGN balance 938706462238359160319928
New contract OUSD balance 15647570848983407793374
```


**Code:**
```const { ethers, getNamedAccounts } = require("hardhat");

async function main() {
  const { governorAddr, deployerAddr, strategistAddr } =
    await getNamedAccounts();
  const sDeployer = await ethers.provider.getSigner(deployerAddr);
  const sGovernor = await ethers.provider.getSigner(governorAddr);

  const cBuyback = await ethers.getContract("Buyback");
  const vault = await ethers.getContract("VaultProxy");
  const cVault = await ethers.getContractAt("VaultAdmin", vault.address);

  const ousdProxy = await ethers.getContract("OUSDProxy");
  const ousd = await ethers.getContractAt("OUSD", ousdProxy.address);
  const cOGN = await ethers.getContractAt(
    [
      "function balanceOf(address owner) external view returns (uint256 balance)",
    ],
    "0x8207c1FfC5B6804F6024322CcF34F29c3541Ae26"
  );

  const oldBuybackAddress = "0x77314EB392b2be47C014cde0706908b3307Ad6a9";

  const oldOgnBalance = await cOGN.balanceOf(oldBuybackAddress);
  const oldOusdBalance = await ousd.balanceOf(oldBuybackAddress);

  const rewardsSource = await cBuyback.rewardsSource();

  const ognBalance = await cOGN.balanceOf(cBuyback.address);
  const ousdBalance = await ousd.balanceOf(cBuyback.address);

  console.log("New Buyback", cBuyback.address);
  console.log("New Buyback governor", await cBuyback.governor());
  console.log("New Buyback RewardsSource", rewardsSource);
  console.log(
    "Updated trustee",
    await cVault.connect(sDeployer).trusteeAddress()
  );
  console.log("Old contract OGN balance", oldOgnBalance.toString());
  console.log("Old contract OUSD balance", oldOusdBalance.toString());
  console.log("New contract OGN balance", ognBalance.toString());
  console.log("New contract OUSD balance", ousdBalance.toString());
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});```

File: contracts/contracts/buyback/Buyback.sol
Comment (DanielVF) [inline contracts/contracts/buyback/Buyback.sol:53]: Turns out we don't actually need this approve to zero here, since OUSD is happy approving to different amounts. 

Also, since this contract is not upgradable, nothing will have run here before setting an approval.

File: contracts/contracts/governance/Strategizable.sol
Comment (DanielVF) [inline contracts/contracts/governance/Strategizable.sol:10]: Let's put some _gap space in this contract, just in case. Good habit to have for inheritable contracts with storage slots.

Comment (shahthepro) [reply contracts/contracts/governance/Strategizable.sol:10]: Done

File: contracts/contracts/buyback/Buyback.sol
Comment (franckc) [inline contracts/contracts/buyback/Buyback.sol:35]: nit: let's document each of the constructor arguments using natspec format https://docs.soliditylang.org/en/v0.8.15/natspec-format.html
I think it's a good habit for any new smart contract code we write...

File: contracts/contracts/buyback/Buyback.sol
Comment (franckc) [reply contracts/contracts/buyback/Buyback.sol:53]: @shahthepro can you address this comment?

**Review by franckc — COMMENTED** [review]
Took a pass at the contract changes. Looking good. Will review the migrations this evening...

**Review by franckc — APPROVED** [review]
Added a few comments. Please take a look before merging. Otherwise LGTM.
Nice work!

File: contracts/storageLayout/mainnet/Buyback.json
Comment (shahthepro) [reply contracts/storageLayout/mainnet/Buyback.json:1]: This was intentionally deleted. BuyBack contract isn't upgradable. So, there's no need to retain the storage slots. The new contract differs from the old one very much (since there are no Oracles and stuffs). Had to delete this because the tests were failing since there's a change in storage slots used by the new contract. This should get added when deploying the new contract though

File: contracts/contracts/buyback/Buyback.sol
Comment (shahthepro) [reply contracts/contracts/buyback/Buyback.sol:53]: Yep, it isn't needed in the constructor. But might still need in the `setUniswapAddr()` function. So, I'm removing this from the constructor but leaving that method as is.

File: contracts/contracts/buyback/Buyback.sol
Comment (shahthepro) [reply contracts/contracts/buyback/Buyback.sol:44]: Done

File: contracts/storageLayout/mainnet/Buyback.json
Comment (DanielVF) [reply contracts/storageLayout/mainnet/Buyback.json:1]: There's an arg you can set to skip the storage slot checks during a single contract deploy.

We've used in the past when doing weird things to the OUSD token.

https://github.com/OriginProtocol/origin-dollar/blob/f0a2ce23663d32a8734d8f90b8f903ff66402961/contracts/utils/deploy.js#L45-L49

**Comment by DanielVF** [comment]
# Code Review


#### Requirements

This change to the buyback contract is to handles using to OGV instead of OGN, and allows buybacks to work without an oracle. 

Funds flow: OUSD is expected to flow in from the vault, and OGV flow from here to the rewards contract.

#### Deployment Considerations

We'll need to hold a vote at some time in the future on what to do with the OGV / OGN left in the old contract.

With the small amounts of OGV liquidity, the initial trades may be inefficient. Will probably break them into multiple.

#### Internal State

_All internal storge state is configuration, outside reentrancy protection._

External state used is the token balances and the uniswap system.

#### Consequences

If this contract were to have an issue, 10% of OUSD yield for a week or two could be stolen. Currently this amounts around $2,000.

#### Possible Attacks

- Misconfiguration at deployment
- MEV / Slippage
- Ownership problems, misconfiguration

#### Logic

- 🟠 Does not fire stratigist set event during contract creation. It's always good to have a full record via events of these config changes.
- 🟠 Similarly UniswapUpdated is not set when the uniswap address is set to zero to disable swaps.
- 🟠 The expiration time for the swap can just be the current timestamp, since this is executing atomicly inside a transaction.

#### Tests

  - [x] Each publicly callable method has a test
  - [x] Each logical branch has a test
  - [ ] Each require() has a test (_"Invalid minExpected value" missing_)
  - [x] Edge conditions are tested

#### Flavor

_The flavor is good. Simple contract with no state._

#### Overflow

- [x] Uses v8 or, Never use "+" or "-", always use safe math
- [x] Check that all for loops use uint256 (_no loops_)

#### Black magic

- [x] Does not contain `selfdestruct`
- [x] Does not use `delegatecall` outside of proxying
- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.

#### Authentication

- [x] Never use tx.origin
- [x] Check that every external/public function should actually be external
- [x] Check that every external/public function has the correct authentication

#### Cryptographic code

_No Crypto code._

#### Gas problems

_NA_

#### External calls

- [x] Contract addresses passed in are validated
- [x] Unsafe external calls
- [x] Reentrancy guards on all internal state changing functions
- [x] Malicious behaviors
- [x] Could fail from stack depth problems (low level calls must require success)
- [x] No slippage attacks (we need to validate expected tokens received)
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Don't call balanceOf for external contracts to determine what they will do, when they instead use internal accounting?

#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [x] Contract has no payable methods.

File: contracts/storageLayout/mainnet/Buyback.json
Comment (shahthepro) [reply contracts/storageLayout/mainnet/Buyback.json]: Have added back this file with the `skipUpgradeSafety` set to `false` in the deployment file

**Comment by shahthepro** [comment]
@DanielVF Thanks for the review, have addressed those comments too

**Review by DanielVF — COMMENTED** [review]
Changes look good. I'll just need to do fork testing and this should be ready to go.

**Comment by DanielVF** [comment]
My fork tests

```python
from world import *

unlock(USDT_BAGS)
unlock(STRATEGIST)
accounts[0].transfer(STRATEGIST, 1e18)
buyback = Contract.from_abi("Buyback", vault_core.trusteeAddress(), buyback.abi)

usdt.transfer(vault_core, 1e6*1e5, {'from': USDT_BAGS})
vault_core.rebase({'from': USDT_BAGS})


buyback.swapNow(1e18, 1, {'from': STRATEGIST, 'gas': 1e6, 'allow_revert': True})

#should fail
buyback.swapNow(1000e18, 200000e18, {'from': STRATEGIST, 'gas': 1e6, 'allow_revert': True})

buyback.swapNow(1000e18, 200000e18, {'from': STRATEGIST, 'gas': 1e6, 'allow_revert': True})
buyback.swapNow(1000e18, 200000e18, {'from': STRATEGIST, 'gas': 1e6, 'allow_revert': True})
buyback.swapNow(1000e18, 200000e18, {'from': STRATEGIST, 'gas': 1e6, 'allow_revert': True})
buyback.swapNow(1000e18, 200000e18, {'from': STRATEGIST, 'gas': 1e6, 'allow_revert': True})
```

**Comment by DanielVF** [comment]
@shahthepro one last change.

Currently we are using low level calls to send the swap command to uniswap. This is because in the past we cared about ignoring reverts on these calls. However in swapNow, we don't care anymore, and can use the high level solidity calls. This adds the additional solidity checks. While I think we have the current version covered from a security point of view, going high level will let the actual error messages from the uniswap contract bubble up, instead of always getting a generic "buyback failed" message.  Will also simplify the contract.

**Comment by shahthepro** [comment]
@DanielVF That sounds good, will make that change

**Comment by DanielVF** [comment]
@shahthepro Bug in change. :)

**Comment by shahthepro** [comment]
> 

Yeah, that was a blind commit, I'm still testing it on fork

**Comment by shahthepro** [comment]
@DanielVF It should be fixed now, works fine on fork too

**Comment by DanielVF** [comment]
Thanks!

**Review by DanielVF — APPROVED** [review]
Thanks @shahthepro. This looks ready to deploy!

**Pull request merged**
